File: examples/go/case08_test.go
package examplesgo

import "testing"

func TestCase08RecoveryBoundary(t *testing.T) {
	weakStore := &RecoveryStore{}
	before := &WeakRecoveryProjection{store: weakStore}
	if !before.Apply(RecoveryRecord{"SHIPPED", 44}) || before.Apply(RecoveryRecord{"PROCESSING", 42}) {
		t.Fatal("the live guard should reject revision 42")
	}
	if weakStore.record.State != "SHIPPED" {
		t.Fatal("revision 44 value should survive")
	}
	after := &WeakRecoveryProjection{store: weakStore}
	if !after.Apply(RecoveryRecord{"PROCESSING", 42}) || weakStore.record.State != "PROCESSING" {
		t.Fatal("the memory-only guard is lost and permits regression")
	}

	strongStore := &RecoveryStore{}
	PersistRecoveryRecord(strongStore, RecoveryRecord{"SHIPPED", 44})
	recovered := RecoverStateWithRevision(strongStore)
	if !recovered.ready || recovered.Apply(RecoveryRecord{"PROCESSING", 42}) {
		t.Fatal("recovered revision 44 must reject revision 42")
	}
	if strongStore.record != (RecoveryRecord{"SHIPPED", 44}) {
		t.Fatal("stale event must not change the durable record")
	}
	if !recovered.Apply(RecoveryRecord{"DELIVERED", 45}) || strongStore.record != (RecoveryRecord{"DELIVERED", 45}) {
		t.Fatal("revision 45 must advance state and revision together")
	}

	rebuilt := RebuildBeforeReady([]RecoveryRecord{{"PROCESSING", 42}, {"SHIPPED", 44}})
	if !rebuilt.ready || rebuilt.Apply(RecoveryRecord{"PROCESSING", 42}) {
		t.Fatal("complete replay must rebuild the revision before readiness")
	}
}

File: examples/java/Case08Test.java
import java.util.List;

public final class Case08Test {
    public static void main(String[] args) {
        Case08.Store weakStore = new Case08.Store();
        Case08.WeakProjection before = new Case08.WeakProjection(weakStore);
        if (!before.apply(new Case08.Record("SHIPPED", 44))
                || before.apply(new Case08.Record("PROCESSING", 42))) {
            throw new AssertionError("live guard must reject revision 42");
        }
        Case08.WeakProjection after = new Case08.WeakProjection(weakStore);
        if (!after.apply(new Case08.Record("PROCESSING", 42))
                || !weakStore.read().state.equals("PROCESSING")) {
            throw new AssertionError("restart loses the volatile guard");
        }

        Case08.Store strongStore = new Case08.Store();
        Case08.persistStateWithRevision(strongStore, new Case08.Record("SHIPPED", 44));
        Case08.StrongProjection recovered = Case08.StrongProjection.recover(strongStore);
        if (!recovered.ready || recovered.apply(new Case08.Record("PROCESSING", 42))
                || !strongStore.read().state.equals("SHIPPED")) {
            throw new AssertionError("recovered revision 44 must reject revision 42");
        }
        if (!recovered.apply(new Case08.Record("DELIVERED", 45))
                || strongStore.read().revision != 45) {
            throw new AssertionError("revision 45 must advance state and revision");
        }

        Case08.StrongProjection rebuilt = Case08.rebuildBeforeReady(List.of(
            new Case08.Record("PROCESSING", 42), new Case08.Record("SHIPPED", 44)));
        if (rebuilt.apply(new Case08.Record("PROCESSING", 42))) {
            throw new AssertionError("replay must rebuild guard before readiness");
        }
    }
}
